src/dbnode/storage/index.go

```diff
@@ -475,17 +498,213 @@ func (i *nsIndex) Tick(c context.Cancellable) (namespaceIndexTickResult, error)
 	return result, multiErr.FinalError()
 }
 
-func (i *nsIndex) ReplaceBlock(
-	blockStart time.Time,
-	segments []segment.Segment,
+func (i *nsIndex) Flush(
+	flush persist.IndexFlush,
+	shards []databaseShard,
 ) error {
 	i.state.RLock()
-	defer i.state.RUnlock()
 	if !i.isOpenWithRLock() {
+		i.state.RUnlock()
 		return errDbIndexUnableToFlushClosed
 	}
 
-	return fmt.Errorf("not implemented")
+	flushable := make([]index.Block, 0, len(i.state.blocksByTime))
+	for _, block := range i.state.blocksByTime {
+		if !i.canFlushBlock(block, shards) {
+			continue
+		}
+		flushable = append(flushable, block)
+	}
```

Comment:
nit: We could make this a sub-function which would make it a little easier to follow the RUnlock()? 

i.e. 
Rlock()
if !open {
  Runlock()
  return err...
} 
flushable := i.flushableBlocksWithRLock()
Runlock()
```

Reply:
sure thing

Reply:
done